Dockerfile: variant support in frozen-images stage

crazy-max opened the pull request:
closes #44005
related #43529
~needs #44034 (for testing)~

**- What I did**

using `TARGETVARIANT` in frozen-images stage implies changes in `download-frozen-image-v2.sh` script to add support for variants so we are able to build against more platforms.

looking at others issues in this repo, this script seems to be used outside the Dockerfile. I didn't  check this case. If it's relevant to also check that then I think using scopeo in the Dockerfile is a better alternative: https://github.com/moby/moby/pull/44005

**- How I did it**

**- How to verify it**

```
$ docker buildx build --platform linux/arm/v7 --target frozen-images .
```

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah commented on contrib/download-frozen-image-v2.sh:
```diff
@@ -310,12 +325,16 @@ while [ $# -gt 0 ]; do
 					unset IFS
 
 					found=""
-					targetArch="$(get_target_arch)"
+					targetArch=$(get_target_arch)
```
Any reason the quotes are removed here?

thaJeztah commented on contrib/download-frozen-image-v2.sh:
```diff
@@ -205,62 +205,77 @@ get_target_arch() {
-				echo "${uArch}"
+			"mips64le")
+				echo mips64le
+				return 0
```
Feels like all of these _literal_ matches would already have been covered by the `*)`. I realise including them all makes it more explicit, but wondering if it's really needed (things would fail either way if it's a bogus architecture)

crazy-max replied:
I would be ok if this script was exclusively used in the Dockerfile. Maybe @cpuguy83 or @tianon could tell.

crazy-max committed:
Dockerfile: variant support in frozen-images stage

using TARGETVARIANT in frozen-images stage implies changes in
`download-frozen-image-v2.sh` script to add support for variants
so we are able to build against more platforms.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

tianon commented:
The problem with using `uname -m` for platform detection in general is that it detects the architecture of the _kernel_, not the architecture of userspace (which is normally what we're actually after).

corhere commented:
> The problem with using `uname -m` for platform detection in general is that it detects the architecture of the _kernel_, not the architecture of userspace (which is normally what we're actually after).

I've looked, but I could not find any distro-independent ways to enumerate the supported ABI personalities of a Linux system. Not even `setarch(8)` from util-linux has that ability. (`setarch --list` [just dumps the static list compiled into the command without querying whether the kernel supports any of them.](https://github.com/util-linux/util-linux/blob/fb34edff70665f6c2fde319fd7a71d7953ce74e3/sys-utils/setarch.c#L129-L245)) While `uname -m` may not be perfect, is it a blocker? The `$TARGETARCH` escape hatch can be used when necessary and we could always improve upon our target arch autodetection later.

tianon commented:
I guess what I'm trying to say is that I don't understand why this PR touches the (already working) `get_target_arch` function at all.  It doesn't seem relevant to adding support for variants to the script.

corhere reviewed (changes requested):
> I guess what I'm trying to say is that I don't understand why this PR touches the (already working) `get_target_arch` function at all. It doesn't seem relevant to adding support for variants to the script.

Good point! Now that you've pointed it out, I wholeheartedly agree. The modifications to the `get_target_arch` function are independent and should be split off into a separate PR.

corhere commented on contrib/download-frozen-image-v2.sh:
```diff
@@ -332,7 +343,7 @@ while [ $# -gt 0 ]; do
 					done
 					if [ -z "$found" ]; then
-						echo >&2 "error: manifest for $maniArch is not found"
+						echo >&2 "error: manifest for $targetArchVariant is not found"
```
[Shell parameter expansion](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html) could be used instead of the `targetArchVariant` logic.
```suggestion
						echo >&2 "error: manifest for ${targetArch}${targetVariant:+/${targetVariant}} is not found"
```

corhere reviewed (approved):
I have a couple of non-blocking nitpicks

crazy-max commented on contrib/download-frozen-image-v2.sh:
```diff
@@ -261,6 +261,13 @@ get_target_arch() {
+		echo "${TARGETVARIANT}"
+		return 0
+	fi
+}
```
To be aligned with the `arch` one if in the future we want to consider detecting the variant

thaJeztah reviewed (approved):
LGTM
